h2/connection.py

```diff
@@ -1646,7 +1646,19 @@ def _receive_data_frame(self, frame):
         self._inbound_flow_control_window_manager.window_consumed(
             flow_controlled_length
         )
-        stream = self._get_stream_by_id(frame.stream_id)
+
+        try:
+            stream = self._get_stream_by_id(frame.stream_id)
```

Comment:
There are two ways to know that stream is closed. First one is to handle a `StreamClosedError`. Second one is to check `if stream.closed`. I think that this issue is valid for both cases.

Reply:
I'm not sure I understand your comment - do you propose a change?

Reply:
Yes. In a case when:
- stream is not yet removed from the `Connection.streams` dict
- `StreamClosedError` is not raised
- `WINDOW_UPDATE` isn't sent
- stream's state machine raises `StreamClosedError` later in `H2StreamStateMachine.recv_on_closed_stream()`
- so only `RST_STREAM` is sent in this case.

I propose to do one of two things:
- handle `StreamClosedError` also during `stream.receive_data(...)`
- or check that stream is not in "open" or "half-closed (local)" states after first try..except block

... and emit `WINDOW_UPDATE` in any of those cases

Reply:
Can `stream.receive_data` even raise a `StreamClosedError`? Can you point me to a possible code path? All state transitions are either valid, or result in `H2StreamStateMachine.reset_stream_on_error` - which we need to handle as discussed.